### PR TITLE
Make use of webpack-cli in webpack

### PR DESCRIPTION
**Pull request by ev1stensberg**
@sokra this is ready for a final review now. @shellscape needs to verify it doesn’t break the dev-server

### CODE REVIEW

**Comment by ev1stensberg** [comment]
Facebook aren't willing to upgrade `jscodeshift` to get rid of the warning ( see: https://github.com/facebook/jscodeshift/pull/198#issuecomment-304329657 )

**Comment by shellscape** [comment]
@ev1stensberg i'll prob have a chance to check this out on 10/22 EDT. the one big thing we'll have to look for is the options validation stuff, which WDS piggybacks on.

**Comment by ev1stensberg** [comment]
If you've hardcoded some logic in the dev-server, the CLI has a lot of the same logic where it used to be @shellscape

**Comment by ev1stensberg** [comment]
@sokra any idea why the tests are failing? Might be https://github.com/webpack/webpack-cli/blob/master/bin/webpack.js#L28-L32

**Comment by ev1stensberg** [comment]
skjermbilde 2017-10-27 kl 00 09 13 weird..

**Comment by ev1stensberg** [comment]
@sokra Tests should run fine now, added a `.nsprc` to avoid breaking the build every time. 

- https://github.com/nodesecurity/nsp#exceptions

- https://nodesecurity.io/advisories/118

**Comment by ev1stensberg** [comment]
@shellscape https://github.com/webpack/webpack-dev-server/pull/1163 

Anything else that needs attention?

File: .nsprc [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "exceptions": ["https://nodesecurity.io/advisories/118"]
```
Comment (sokra) [inline .nsprc]: Is this fixes? Could you remove this file?

Comment (ev1stensberg) [reply .nsprc]: it will make the nsp check succeed

**Comment by ev1stensberg** [comment]
@sokra what do you wanna do about the security issue? Adding a nsprc will make the CI pass

**Comment by sokra** [comment]
https://twitter.com/wSokra/status/928885927803580416

**Comment by sokra** [comment]
https://github.com/facebook/jscodeshift/issues/229

**Comment by sokra** [comment]
test failure because of #5359

**Comment by ev1stensberg** [comment]
I'll take a look at it tomorrow and get this PR updated

**Comment by ev1stensberg** [comment]
@sokra  CLI re-synced with upstream branch ( next ), rebased this PR against the next branch as well. What should we do with the facebook issue?

**Comment by sokra** [comment]
@ev1stensberg facebook issue is in progress.

Could you update this to the next branch again?

I changed a few things in the CLI on this branch:

* added a `--mode` argument -> `mode` configuration option
* `-p` sets `--mode` to `production`
* `-d` sets `--mode` to `development`
* `-p` no longer defines `NODE_ENV` (this is now done by the `mode` option)
* `-d` no longer sets the devtool (this is now done by the `mode` option)

**Comment by ev1stensberg** [comment]
fyi @sokra Got an exam period right now, so earliest would be on Wednesday

**Comment by sokra** [comment]
No hurry...

**Comment by ev1stensberg** [comment]
@sokra Could you point me to the PR this was implemented in?

**Comment by sokra** [comment]
https://github.com/webpack/webpack/pull/5994

**Comment by ev1stensberg** [comment]
@sokra CLI is resynced against the next branch again. Before merging, any potential changes to `binTestCases` or the `/bin` folder, could you also redirect them to the CLI repo, or chill them until the CLI gets merged? Avoids extra refactoring work

**Comment by sokra** [comment]
> Before merging, any potential changes to binTestCases or the /bin folder, could you also redirect them to the CLI repo, or chill them until the CLI gets merged? Avoids extra refactoring work

I try to.

**Comment by sokra** [comment]
After a long fight js-codeshift was updated...

So we can merge this PR once dependencies are updated in webpack-cli resp. lockfiles in this PR.

**Comment by ev1stensberg** [comment]
CI build failed on yarn/cache, but I think this is ready now @sokra

**Comment by sokra** [comment]
Awesome ❤️

**Pull request merged**